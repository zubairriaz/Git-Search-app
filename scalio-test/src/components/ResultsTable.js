/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";
import React, { useEffect, useState } from "react";
import arrowdown from "../static/images/arrow-down.png";
import arrowup from "../static/images/arrow-up.png";

export default function ResultsTable({ columns, data }) {
	const [items, setItems] = useState([]);
	const [stateColumns, setStateColumns] = useState([]);

	useEffect(() => {
		let column = columns.filter((column) => column.isSorted)[0];
		let sortedData = sortData(data, column, false);
		setItems(sortedData);
		setStateColumns(columns);
	}, []);

	useEffect(() => {
		let column = columns.filter((column) => column.isSorted)[0];
		let sortedData = sortData(data, column, false);
		setItems(sortedData);
	}, [data]);

	const SortByColumn = (column) => {
		let sortedData, updatedColumns;
		let col = stateColumns.filter((column) => column.isSorted)[0];
		if (col.key == column.key) {
			sortedData = sortData(data, column, toggle());
			updatedColumns = updateColumns(stateColumns, column);
		} else {
			sortedData = sortData(data, column, false);
			updatedColumns = updateColumns(stateColumns, column, false);
		}
		setItems([...sortedData]);
		setStateColumns([...updatedColumns]);
	};

	const TableHeader = ({ columns }) => {
		return (
			<thead>
				<tr css={getTrCss()}>
					{columns.map((column) => (
						<th
							css={css`
								padding-right: 100px;
								border: 1px solid black;
							`}
							key={column.id}
						>
							<div>
								<span
									css={getHeaderSpanCss()}
									onClick={() => SortByColumn(column)}
								>
									{column.name}

									{column &&
										column.isSorted &&
										!column.isSortedAgain && (
											<img src={arrowdown}></img>
										)}
									{column &&
										column.isSorted &&
										column.isSortedAgain && (
											<img src={arrowup}></img>
										)}
								</span>
							</div>
						</th>
					))}
				</tr>
			</thead>
		);
	};

	return (
		<React.Fragment>
			<table css={getTableCss()}>
				<TableHeader columns={stateColumns}></TableHeader>
				<TableRows data={items} columns={stateColumns}></TableRows>
			</table>
		</React.Fragment>
	);
}

const TableRow = ({ item, columns }) => {
	return (
		<tr>
			{columns.map((column) => {
				if (column.onRender) {
					return (
						<td key={column.id}>
							<span>{column.onRender(item)}</span>
						</td>
					);
				}
				return <td key={column.id}><span>{item[column.key]}</span></td>;
			})}
		</tr>
	);
};

const TableRows = ({ data, columns }) => {
	return (
		<tbody>
			{data.map((item) => {
				return (
					<TableRow
						key={item.id}
						item={item}
						columns={columns}
					></TableRow>
				);
			})}
		</tbody>
	);
};

const getTrCss = () => {
	return css`
		border: 1px solid gray;
	`;
};

const getTableCss = () => {
	return css`
		box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
		transition: 0.3s;
		width: 100%;
		padding: 10px 30px;
		font-size: 14px;
		border-radius: 4px;
	`;
};

const getHeaderSpanCss = () => {
	return css`
		color: gray;
		font-size: 16px;
		font-weight: 400;
		float: left;
		&:hover {
			cursor: pointer;
		}
	`;
};

const sortObj = (property) => (a, b) => {
	return a[property].localeCompare(b[property]);
};

const invert =
	(fn) =>
	(...args) =>
		-fn(...args);

const toggle = (function () {
	let bit = false;
	return () => {
		bit = !bit;
		return bit;
	};
})();

const sortData = (data, column, reverse = false) => {
	let sortedData;
	data = copy(data);
	if (reverse) {
		sortedData = data.sort(invert(sortObj(column["key"])));
	} else {
		sortedData = data.sort(sortObj(column["key"]));
	}
	return sortedData;
};

const updateColumns = (columns, column, again = true) => {
	return columns.map((item) =>
		item.key == column.key
			? again
				? {
						...item,
						isSorted: true,
						isSortedAgain: !item.isSortedAgain,
				  }
				: {
						...item,
						isSorted: true,
						isSortedAgain: false,
				  }
			: { ...item, isSorted: false, isSortedAgain: false }
	);
};

const copy = (data) => {
	return data ? JSON.parse(JSON.stringify(data)) : undefined;
};
