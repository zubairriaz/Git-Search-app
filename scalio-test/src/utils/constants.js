export const url =
	"https://api.github.com/search/users?q={login} in:login&per_page=9";
export const SET_ITEM = "APP/ADD_ITEMS";
export const SET_ISLOADING = "APP/IS_LOADING";
export const SET_ERROR_MESSAGE = "APP/SET_ERROR_MESSAGE";

export const PAGE_LIMT = 9;

export const columns = [
	{
		id: 1,
		key: "avatar_url",
        name:"Avatar",
        isSorted:false,
        isSortedAgain:false,
		onRender: (item) => {
			return (
				<img
                    width="50px"
                    height="50px"
					src={item["avatar_url"]}
					alt="avatar_url"
				></img>
			);
		},
	},
	{ id: 2, key: "login",name:"Login", isSorted: true,isSortedAgain:false },
	{ id: 3, name:"Type",key: "type",isSorted:false,isSortedAgain:false },
];
