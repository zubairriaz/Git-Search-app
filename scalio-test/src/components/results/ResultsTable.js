/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";
import React, { useEffect, useState } from "react";
import arrowdown from "../../static/images/arrow-down.png";
import arrowup from "../../static/images/arrow-up.png";
import {sortData,toggle,updateColumns} from "../../utils/utils"
import Pagination from "./pagination";
import { PAGE_LIMT } from "../../utils/constants";

export default function ResultsTable({ columns, data, currentPage, pages, setPage }) {
	const [items, setItems] = useState([]);
	const [stateColumns, setStateColumns] = useState([]);

	useEffect(() => {
		let column = columns.filter((column) => column.isSorted)[0];
		let sortedData = sortData(data, column, false);
		setItems(sortedData);
		setStateColumns(columns);
	}, []);

	useEffect(() => {
		let column = columns.filter((column) => column.isSorted)[0];
		let sortedData = sortData(data, column, false);
		setItems(sortedData);
	}, [columns,data]);

	function SortByColumn(column) {
		let sortedData, updatedColumns;
		let col = stateColumns.filter((column) => column.isSorted)[0];
		if (col.key ===  column.key) {
			sortedData = sortData(data, column, toggle());
			updatedColumns = updateColumns(stateColumns, column);
		} else {
			sortedData = sortData(data, column, false);
			updatedColumns = updateColumns(stateColumns, column, false);
		}
		setItems([...sortedData]);
		setStateColumns([...updatedColumns]);
	};

	const TableHeader = ({ columns }) => {
		return (
			<thead>
				<tr css={getTrCss()}>
					{columns.map((column) => (
						<th
							css={css`
								border: 1px solid lightblue;
							`}
							key={column.id}
						>
							<div>
								<span
									css={getHeaderSpanCss()}
									onClick={() => SortByColumn(column)}
								>
									{column.name}

									{column &&
										column.isSorted &&
										!column.isSortedAgain && (
											<img src={arrowdown}></img>
										)}
									{column &&
										column.isSorted &&
										column.isSortedAgain && (
											<img src={arrowup}></img>
										)}
								</span>
							</div>
						</th>
					))}
				</tr>
			</thead>
		);
	};

	return (
		<React.Fragment>
			<table css={getTableCss()}>
				<TableHeader columns={stateColumns}></TableHeader>
				<TableRows data={items} columns={stateColumns}></TableRows>
			</table>
			<Pagination
						currentPage={currentPage}
						pages={pages}
						pageLimit={PAGE_LIMT}
						setPage={setPage}
					></Pagination>
		</React.Fragment>
	);
}

const TableRow = ({ item, columns }) => {
	return (
		<tr>
			{columns.map((column) => {
				if (column.onRender) {
					return (
						<td key={column.id}>
							<span>{column.onRender(item)}</span>
						</td>
					);
				}
				return <td key={column.id}><span>{item[column.key]}</span></td>;
			})}
		</tr>
	);
};

const TableRows = ({ data, columns }) => {
	return (
		<tbody>
			{data.map((item) => {
				return (
					<TableRow
						key={item.id}
						item={item}
						columns={columns}
					></TableRow>
				);
			})}
		</tbody>
	);
};

const getTrCss = () => {
	return css`
		border: 1px solid gray;
	`;
};

const getTableCss = () => {
	return css`
		box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
		transition: 0.3s;
		width: 100%;
		padding: 10px 30px;
		font-size: 14px;
		border-radius: 4px;
	`;
};

const getHeaderSpanCss = () => {
	return css`
		color: gray;
		font-size: 16px;
		font-weight: 400;
		float: left;
		&:hover {
			cursor: pointer;
		}
	`;
};
