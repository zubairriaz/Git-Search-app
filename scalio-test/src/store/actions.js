import {SET_ITEM, SET_ISLOADING, SET_ERROR_MESSAGE} from '../utils/constants'

export const setItems = (item) => ({
    type: SET_ITEM,
    item,
  });

  export const setErrorMessgae = (message) => ({
    type: SET_ERROR_MESSAGE,
    message,
  });

  export const setIsLoading = (bool) => ({
    type: SET_ISLOADING,
    bool,
  });
  
  