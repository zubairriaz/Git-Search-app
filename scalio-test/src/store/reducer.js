import { SET_ITEM, SET_ISLOADING, SET_ERROR_MESSAGE } from "../utils/constants";
export const initialState = {
	item: {},
	isLoading: false,
	errorMessage: "",
};


export const appReducer = (state = initialState, action) => {
	if (action.type === SET_ITEM) {
		return {
			...state,
			item: { ...action.item },
			isLoading: false,
      errorMessage:""
		};
	}
	if (action.type === SET_ERROR_MESSAGE) {
		return {
			...state,
			errorMessage: extractMessage(action.message),
      item:{items:[],total_count:0},
			isLoading: false,
		};
	}
	if (action.type === SET_ISLOADING) {
		return {
			...state,
			isLoading: action.bool,
		};
	}
};


const extractMessage=res=>res? res.data? res.data.message : "" :"";