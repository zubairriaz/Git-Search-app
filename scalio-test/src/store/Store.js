import React,{useContext} from 'react';
import request from '../utils/request';

const Store = React.createContext();

export const useStore = () => useContext(Store);

export const StoreProvider = ({ children, initialState, reducer }) => {
  const [globalState, dispatch] = React.useReducer(reducer, initialState);
  const apiRequest = request(dispatch)

  return (
    <Store.Provider value={[globalState, dispatch, apiRequest]}>{children}</Store.Provider>
  );
};