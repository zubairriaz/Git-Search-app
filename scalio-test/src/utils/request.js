import axios from "axios";
import { setItems,setIsLoading , setErrorMessgae} from "../store/actions";

const extractData = (response) => (response ? response.data : null);

function request(dispatch) {
	return async (url) => {
        try{
        dispatch(setIsLoading(true))
		let response =await axios.get(url);
		dispatch(setItems(extractData(response)));
        }catch(error){
            dispatch(setIsLoading(false))
            let message = error.response
            console.log(error.response)
            dispatch(setErrorMessgae(message))
        }
	};
}

export default request
