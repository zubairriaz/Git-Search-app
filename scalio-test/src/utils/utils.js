import axios from "axios";
import { setItems, setIsLoading, setErrorMessgae } from "../store/actions";

export const extractData = (response) => (response ? response.data : null);

export const request = (dispatch) => {
	return async (url) => {
		try {
			dispatch(setIsLoading(true));
			let response = await axios.get(url);
			dispatch(setItems(extractData(response)));
		} catch (error) {
			dispatch(setIsLoading(false));
			let message = error.response;
			console.log(error.response);
			dispatch(setErrorMessgae(message));
		}
	};
};

export const sortObj = (property) => (a, b) => {
	return a[property].localeCompare(b[property]);
};

export const invert =
	(fn) =>
	(...args) =>
		-fn(...args);

export const toggle = (function () {
	let bit = false;
	return () => {
		bit = !bit;
		return bit;
	};
})();

export const sortData = (data, column, reverse = false) => {
	let sortedData;
	data = copy(data);
	if (reverse) {
		sortedData = data.sort(invert(sortObj(column["key"])));
	} else {
		sortedData = data.sort(sortObj(column["key"]));
	}
	return sortedData;
};

const getObj = (item, toggle) => ({
	...item,
	isSorted: true,
	isSortedAgain: toggle ? !item.isSortedAgain : false,
});

export const updateColumns = (columns, column, again = true) => {
	return columns.map((item) =>
		item.key === column.key
			? again
				? getObj(item, true)
				: getObj(item,false)
			: { ...item, isSorted: false, isSortedAgain: false }
	);
};

export const copy = (data) => {
	return data ? JSON.parse(JSON.stringify(data)) : undefined;
};

export const replaceText = (url, searchTerm) =>
	url.replace("{login}", searchTerm);
