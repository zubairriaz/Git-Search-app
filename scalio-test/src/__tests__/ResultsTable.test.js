import React from 'react';

import { fireEvent, render} from '@testing-library/react';

import { shallow } from 'enzyme';

import ResultsTable from '../components/ResultsTable';
import {StoreProvider} from "../store/Store"
import {columns} from "../utils/constants"

test('Check If Results Table Loads Successfully', () => {
  const search = shallow(<StoreProvider><ResultsTable/></StoreProvider>);
  expect(search).toMatchSnapshot();
});
let table;
beforeEach(()=>{table = render(<StoreProvider><ResultsTable columns={columns} data={[]}/></StoreProvider>)})


test('Table Have First columns',  () => {
   const header = table.getByText(columns[0].name)  
    expect(header).toBeDefined();
});

test('Table Have last Column',  () => {
    const header = table.getByText(columns[columns.length - 1].name)  
     expect(header).toBeDefined();
 });


    

 






