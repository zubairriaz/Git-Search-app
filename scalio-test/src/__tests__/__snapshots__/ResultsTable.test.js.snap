// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check If Results Table Loads Successfully 1`] = `ShallowWrapper {}`;

exports[`Check If Search Loads Successfully 1`] = `ShallowWrapper {}`;
